File: Android/SayHi/app/src/main/java/de/oklemenz/sayhi/service/Analytics.java
package de.oklemenz.sayhi.service;

import android.os.Bundle;
import android.text.TextUtils;

import com.google.firebase.analytics.FirebaseAnalytics;

import de.oklemenz.sayhi.AppDelegate;
import de.oklemenz.sayhi.model.Category;
import de.oklemenz.sayhi.model.Match;
import de.oklemenz.sayhi.model.Profile;
import de.oklemenz.sayhi.model.Settings;
import de.oklemenz.sayhi.model.StageCategory;
import de.oklemenz.sayhi.model.StageTag;
import de.oklemenz.sayhi.model.Tag;
import de.oklemenz.sayhi.model.UserData;

/**
 * Created by Oliver Klemenz on 22.02.17.
 */

public class Analytics {

    private static Analytics instance = new Analytics();

    public static Analytics getInstance() {
        return instance;
    }

    private Bundle parameters(Profile profile) {
        Bundle parameters = new Bundle();
        String matchMode = profile.matchMode != null ? profile.matchMode.toExternalString() : "";
        int matchCode = profile.matchMode != null ? profile.matchMode.code : -1;
        parameters.putString("profileMatchMode", matchMode);
        parameters.putInt("profileMatchCode", matchCode);
        parameters.putString("profileRelationType", profile.relationType.code);
        parameters.putInt("profilePosTagCount", profile.posTags.size());
        parameters.putInt("profileNegTagCount", profile.negTags.size());
        return parameters;
    }

    private Bundle parameters(Tag tag) {
        Bundle parameters = new Bundle();
        parameters.putString("tagKey", tag.key);
        parameters.putString("tagName", tag.getName());
        parameters.putString("tagPrimaryLangKey", tag.primaryLangKey);
        parameters.putString("tagRefKey", tag.refKey != null ? tag.refKey : "");
        parameters.putString("tagRefPrimaryLangKey", tag.getRefPrimaryLangKey() != null ? tag.getRefPrimaryLangKey() : "");
        parameters.putString("tagEffectiveKey", tag.getEffectiveKey());
        parameters.putString("categoryKey", tag.getCategoryKey());
        parameters.putString("categoryName", tag.getCategory() != null ? tag.getCategory().getName() : "");
        parameters.putString("categoryPrimaryLangKey", tag.getCategory() != null ? tag.getCategory().primaryLangKey : "");
        parameters.putString("categoryRefKey", tag.getCategory() != null ? tag.getCategory().refKey : "");
        parameters.putString("categoryRefPrimLangKey", tag.getCategory() != null ? tag.getCategory().getRefPrimaryLangKey() : "");
        return parameters;
    }

    private Bundle parameters(Category category) {
        Bundle parameters = new Bundle();
        parameters.putString("categoryKey", category.key);
        parameters.putString("categoryName", category.getName());
        parameters.putString("categoryPrimaryLangKey", category.primaryLangKey);
        parameters.putString("categoryRefKey", category.refKey != null ? category.refKey : "");
        parameters.putString("categoryRefPrimLangKey", category.getRefPrimaryLangKey() != null ? category.getRefPrimaryLangKey() : "");
        parameters.putString("categoryEffectiveKey", category.getEffectiveKey());
        return parameters;
    }

    private Bundle parameters(StageTag tag) {
        Bundle parameters = new Bundle();
        parameters.putString("tagKey", tag.key);
        parameters.putString("tagName", tag.name);
        parameters.putString("categoryKey", tag.categoryKey);
        parameters.putString("categoryName", tag.categoryName);
        parameters.putLong("counter", tag.counter);
        return parameters;
    }

    private Bundle parameters(StageCategory category) {
        Bundle parameters = new Bundle();
        parameters.putString("categoryKey", category.key);
        parameters.putString("categoryName", category.name);
        parameters.putLong("counter", category.counter);
        return parameters;
    }

    private Bundle parameters(Match match) {
        Bundle parameters = new Bundle();
        parameters.putString("matchDate", Utilities.getISO8601StringForDate(match.date));
        parameters.putString("matchMode", match.mode.toString());
        parameters.putInt("matchCode", match.mode.code);
        parameters.putBoolean("matchHandshake", match.handshake);
        parameters.putString("profileRelationType", match.relationType.code);
        parameters.putInt("profilePosTagCount", match.profilePosTagCount);
        parameters.putInt("profileNegTagCount", match.profileNegTagCount);
        parameters.putString("locationLongitude", match.locationLongitude);
        parameters.putString("locationLatitude", match.locationLatitude);
        parameters.putString("locationStreet", match.locationStreet);
        parameters.putString("locationCity", match.locationCity);
        parameters.putString("locationCountry", match.locationCountry);
        parameters.putString("messageLanguage", match.langCode);
        parameters.putString("messageGender", match.gender.code);
        parameters.putInt("messageBirthYear", match.birthYear);
        parameters.putInt("messageAge", match.age());
        parameters.putString("messageInstallation", match.installationUUID);
        parameters.putInt("messagePosTagCount", match.messagePosTagCount);
        parameters.putInt("messageNegTagCount", match.messageNegTagCount);
        parameters.putString("matchLeftLeftTagKeys", TextUtils.join(",", match.getBothPosTagEffectiveKeys()));
        parameters.putString("matchRightRightTagKeys", TextUtils.join(",", match.getBothNegTagEffectiveKeys()));
        parameters.putString("matchLeftRightTagKeys", TextUtils.join(",", match.getOnlyPosTagEffectiveKeys()));
        parameters.putString("matchRightLeftTagKeys", TextUtils.join(",", match.getOnlyNegTagEffectiveKeys()));
        return parameters;
    }

    public void logFirstStart() {
        Bundle parameters = new Bundle();
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("first_start", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("first_start", parameters);
        }
    }

    public void logAddProfile(Profile profile) {
        Bundle parameters = parameters(profile);
        parameters.putLong(FirebaseAnalytics.Param.VALUE, 1);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("profile", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("profile", parameters);
        }
    }

    public void logChangeProfile(Profile profile) {
        Bundle parameters = parameters(profile);
        parameters.putLong(FirebaseAnalytics.Param.VALUE, 0);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("profile", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("profile", parameters);
        }
    }

    public void logRemoveProfile(Profile profile) {
        Bundle parameters = parameters(profile);
        parameters.putLong(FirebaseAnalytics.Param.VALUE, -1);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("profile", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("profile", parameters);
        }
    }

    public void logPositive(Tag tag, long previousValue) {
        Bundle parameters = parameters(tag);
        parameters.putLong(FirebaseAnalytics.Param.VALUE, 1);
        parameters.putLong("previousValue", previousValue);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("tag_assign", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("tag_assign", parameters);
        }
    }

    public void logNegative(Tag tag, long previousValue) {
        Bundle parameters = parameters(tag);
        parameters.putLong(FirebaseAnalytics.Param.VALUE, -1);
        parameters.putLong("previousValue", previousValue);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("tag_assign", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("tag_assign", parameters);
        }
    }

    public void logNeutral(Tag tag, long previousValue) {
        Bundle parameters = parameters(tag);
        parameters.putLong(FirebaseAnalytics.Param.VALUE, 0);
        parameters.putLong("previousValue", previousValue);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("tag_assign", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("tag_assign", parameters);
        }
    }

    public void logMatch(Match match, String session, String messageSession) {
        Bundle parameters = parameters(match);
        parameters.putString("matchSession", session);
        parameters.putString("messageSession", messageSession);
        parameters.putString("matchLeftLeftTagKeys", Utilities.prefix(parameters.getString("matchLeftLeftTagKeys"), 100));
        parameters.putString("matchRightRightTagKeys", Utilities.prefix(parameters.getString("matchRightRightTagKeys"), 100));
        parameters.putString("matchLeftRightTagKeys", Utilities.prefix(parameters.getString("matchLeftRightTagKeys"), 100));
        parameters.putString("matchRightLeftTagKeys", Utilities.prefix(parameters.getString("matchRightLeftTagKeys"), 100));
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("match", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            parameters = parameters(match);
            DataService.getInstance().logEvent("match", parameters);
        }
    }

    public void logNewTag(StageTag tag) {
        Bundle parameters = parameters(tag);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("tag_new", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("tag_new", parameters);
        }
    }

    public void logNewCategory(StageCategory category) {
        Bundle parameters = parameters(category);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("category_new", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("category_new", parameters);
        }
    }

    public void logInviteFriend() {
        Bundle parameters = new Bundle();
        parameters.putInt("count", UserData.getInstance().inviteFriendSentCount);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("invite_friend", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("invite_friend", parameters);
        }
    }

    public void logRateApp() {
        Bundle parameters = new Bundle();
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("rate_app", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("rate_app", parameters);
        }
    }

    public void logSupportMail() {
        Bundle parameters = new Bundle();
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("support_feedback", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("support_feedback", parameters);
        }
    }

    public void logRecordVoice() {
        Bundle parameters = new Bundle();
        parameters.putLong(FirebaseAnalytics.Param.VALUE, 1);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("record_voice", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("record_voice", parameters);
        }
    }

    public void logRerecordVoice() {
        Bundle parameters = new Bundle();
        parameters.putLong(FirebaseAnalytics.Param.VALUE, 0);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("record_voice", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("record_voice", parameters);
        }
    }

    public void logRemoveRecordedVoice() {
        Bundle parameters = new Bundle();
        parameters.putLong(FirebaseAnalytics.Param.VALUE, -1);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("record_voice", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("record_voice", parameters);
        }
    }

    public void logSpaceSwitched(String space) {
        Bundle parameters = new Bundle();
        parameters.putString("space", space);
        if (!Settings.getInstance().getDisableRecordAnalytics()) {
            FirebaseAnalytics.getInstance(AppDelegate.getInstance().Context).logEvent("space_switch", parameters);
        }
        if (!Settings.getInstance().getDisableRecordAnalyticsDB()) {
            DataService.getInstance().logEvent("space_switch", parameters);
        }
    }
}

File: iOS/SayHi/SayHi/Analytics.swift
//
//  Analytics.swift
//  SayHi
//
//  Created by Klemenz, Oliver on 21.11.16.
//  Copyright © 2016 Klemenz, Oliver. All rights reserved.
//

import Foundation
import FirebaseAnalytics

func +<K, V> ( left: [K : V], right: [K : V]) -> [K : V] {
    var map = [K : V]()
    for (k, v) in left {
        map[k] = v
    }
    for (k, v) in right {
        map[k] = v
    }
    return map
}

class Analytics {
    
    static let instance = Analytics()

    func parameters(_ profile : Profile) -> [String:NSObject] {
        let matchMode = profile.matchMode != nil ? profile.matchMode!.externalDescription : ""
        let matchCode = profile.matchMode != nil ? profile.matchMode!.rawValue : -1
        return [
            "profileMatchMode" : matchMode as NSObject,
            "profileMatchCode" : matchCode as NSObject,
            "profileRelationType" : profile.relationType.rawValue as NSObject,
            "profilePosTagCount" : profile.posTags.count as NSObject,
            "profileNegTagCount" : profile.negTags.count as NSObject
        ]
    }
    
    func parameters(_ tag : Tag) -> [String:NSObject] {
        return [
            "tagKey" : tag.key as NSObject,
            "tagName" : tag.name as NSObject,
            "tagPrimaryLangKey" : tag.primaryLangKey as NSObject,
            "tagRefKey" : (tag.refKey ?? "") as NSObject,
            "tagRefPrimaryLangKey" : (tag.refPrimaryLangKey ?? "") as NSObject,
            "tagEffectiveKey" : tag.effectiveKey as NSObject,
            "categoryKey" : tag.categoryKey as NSObject,
            "categoryName" : (tag.category?.name ?? "") as NSObject,
            "categoryPrimaryLangKey" : (tag.category?.primaryLangKey ?? "") as NSObject,
            "categoryRefKey" : (tag.category?.refKey ?? "") as NSObject,
            "categoryRefPrimLangKey" : (tag.category?.refPrimaryLangKey ?? "") as NSObject
        ]
    }
    
    func parameters(_ category : Category) -> [String:NSObject] {
        return [
            "categoryKey" : category.key as NSObject,
            "categoryName" : category.name as NSObject,
            "categoryPrimaryLangKey" : category.primaryLangKey as NSObject,
            "categoryRefKey" : (category.refKey ?? "") as NSObject,
            "categoryRefPrimLangKey" : (category.refPrimaryLangKey ?? "") as NSObject,
            "categoryEffectiveKey" : category.effectiveKey as NSObject
        ]
    }
    
    func parameters(_ tag : StageTag) -> [String:NSObject] {
        return [
            "tagKey" : tag.key as NSObject,
            "tagName" : tag.name as NSObject,
            "categoryKey" : tag.categoryKey as NSObject,
            "categoryName" : tag.categoryName as NSObject,
            "counter" : tag.counter as NSObject
        ]
    }
    
    func parameters(_ category : StageCategory) -> [String:NSObject] {
        return [    
            "categoryKey" : category.key as NSObject,
            "categoryName" : category.name as NSObject,
            "counter" : category.counter as NSObject
        ]
    }
    
    func parameters(_ match : Match) -> [String:NSObject] {
        return [
            "matchDate" : match.date.iso as NSObject,
            "matchMode" : match.mode.description as NSObject,
            "matchCode" : match.mode.rawValue as NSObject,
            "matchHandshake" : match.handshake as NSObject,
            "profileRelationType" : match.relationType.rawValue as NSObject,
            "profilePosTagCount" : match.profilePosTagCount as NSObject,
            "profileNegTagCount" : match.profileNegTagCount as NSObject,
            "locationLongitude" : match.locationLongitude as NSObject,
            "locationLatitude" : match.locationLatitude as NSObject,
            "locationStreet" : match.locationStreet as NSObject,
            "locationCity" : match.locationCity as NSObject,
            "locationCountry" : match.locationCountry as NSObject,
            "messageLanguage" : match.langCode as NSObject,
            "messageGender" : match.gender.rawValue as NSObject,
            "messageBirthYear" : match.birthYear as NSObject,
            "messageAge" : match.age as NSObject,
            "messageInstallation" : match.installationUUID as NSObject,
            "messagePosTagCount" : match.messagePosTagCount as NSObject,
            "messageNegTagCount" : match.messageNegTagCount as NSObject,
            "matchLeftLeftTagKeys" : match.bothPosTagEffectiveKeys.joined(separator: ",") as NSObject,
            "matchRightRightTagKeys" : match.bothNegTagEffectiveKeys.joined(separator: ",") as NSObject,
            "matchLeftRightTagKeys" : match.onlyPosTagEffectiveKeys.joined(separator: ",") as NSObject,
            "matchRightLeftTagKeys" : match.onlyNegTagEffectiveKeys.joined(separator: ",") as NSObject
        ]
    }
    
    func logFirstStart()  {
        let params: [String:NSObject] = [:]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("first_start", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("first_start", parameters: params)
        }
    }
    
    func logAddProfile(profile : Profile)  {
        let params = parameters(profile) + [
            AnalyticsParameterValue : 1 as NSObject,
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("profile", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("profile", parameters: params)
        }
    }
    
    func logChangeProfile(profile : Profile)  {
        let params = parameters(profile) + [
            AnalyticsParameterValue : 0 as NSObject,
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("profile", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("profile", parameters: params)
        }
    }
    
    func logRemoveProfile(profile : Profile)  {
        let params = parameters(profile) + [
            AnalyticsParameterValue : -1 as NSObject,
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("profile", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("profile", parameters: params)
        }
    }
    
    func logPositive(tag : Tag, previousValue : Int) {
        let params = parameters(tag) + [
            AnalyticsParameterValue : 1 as NSObject,
            "previousValue": previousValue as NSObject
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("tag_assign", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("tag_assign", parameters: params)
        }
    }
    
    func logNegative(tag : Tag, previousValue : Int) {
        let params = parameters(tag) + [
            AnalyticsParameterValue : -1 as NSObject,
            "previousValue": previousValue as NSObject
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("tag_assign", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("tag_assign", parameters: params)
        }
    }
    
    func logNeutral(tag : Tag, previousValue : Int) {
        let params = parameters(tag) + [
            AnalyticsParameterValue : 0 as NSObject,
            "previousValue": previousValue as NSObject
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("tag_assign", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("tag_assign", parameters: params)
        }
    }
    
    func logMatch(match: Match, session: String, messageSession: String) {
        var params = parameters(match)
        params["matchSession"] = session as NSObject
        params["messageSession"] = messageSession as NSObject
        params["matchLeftLeftTagKeys"] = (params["matchLeftLeftTagKeys"] as! String).prefix(100) as NSObject
        params["matchRightRightTagKeys"] = (params["matchRightRightTagKeys"] as! String).prefix(100) as NSObject
        params["matchLeftRightTagKeys"] = (params["matchLeftRightTagKeys"] as! String).prefix(100) as NSObject
        params["matchRightLeftTagKeys"] = (params["matchRightLeftTagKeys"] as! String).prefix(100) as NSObject
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("match", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            let params = parameters(match)
            DataService.instance.logEvent("match", parameters: params)
        }
    }
    
    func logNewTag(tag : StageTag) {
        let params = parameters(tag)
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("tag_new", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("tag_new", parameters: params)
        }
    }
    
    func logNewCategory(category : StageCategory) {
        let params = parameters(category)
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("category_new", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("category_new", parameters: params)
        }
    }
    
    func logInviteFriend() {
        let params = [
            "count" : UserData.instance.inviteFriendSentCount as NSObject
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("invite_friend", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("invite_friend", parameters: params)
        }
    }
    
    func logRateApp() {
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("rate_app", parameters: [:])
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("rate_app", parameters: [:])
        }
    }
    
    func logSupportMail() {
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("support_feedback", parameters: [:])
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("support_feedback", parameters: [:])
        }
    }
    
    func logRecordVoice() {
        let params = [
            AnalyticsParameterValue : 1 as NSObject,
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("record_voice", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("record_voice", parameters: params)
        }
    }
    
    func logRerecordVoice() {
        let params = [
            AnalyticsParameterValue : 0 as NSObject,
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("record_voice", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("record_voice", parameters: params)
        }
    }
    
    func logRemoveRecordedVoice() {
        let params = [
            AnalyticsParameterValue : -1 as NSObject,
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("record_voice", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("record_voice", parameters: params)
        }
    }
    
    func logSpaceSwitched(_ space: String) {
        let params = [
            "space" : space as NSObject,
        ]
        if !Settings.instance.disableRecordAnalytics {
            FirebaseAnalytics.Analytics.logEvent("space_switch", parameters: params)
        }
        if !Settings.instance.disableRecordAnalyticsDB {
            DataService.instance.logEvent("space_switch", parameters: params)
        }
    }
}
